src/tools/compiletest/src/read2.rs

```diff
@@ -74,6 +27,87 @@ pub fn read2_abbreviated(mut child: Child) -> io::Result<Output> {
     Ok(Output { status, stdout: stdout.into_bytes(), stderr: stderr.into_bytes() })
 }
 
+const HEAD_LEN: usize = 160 * 1024;
+const TAIL_LEN: usize = 256 * 1024;
+const EXCLUDED_PLACEHOLDER_LEN: isize = 32;
+
+enum ProcOutput {
+    Full { bytes: Vec<u8>, excluded_len: isize },
+    Abbreviated { head: Vec<u8>, skipped: usize, tail: Box<[u8]> },
+}
+
+impl ProcOutput {
+    fn new() -> Self {
+        ProcOutput::Full { bytes: Vec::new(), excluded_len: 0 }
+    }
+
+    fn extend(&mut self, data: &[u8], exclude_from_len: &[String]) {
+        let new_self = match *self {
+            ProcOutput::Full { ref mut bytes, ref mut excluded_len } => {
+                let old_len = bytes.len();
+                bytes.extend_from_slice(data);
+
+                // We had problems in the past with tests failing only in some environments,
+                // due to the length of the base path pushing the output size over the limit.
+                //
+                // To make those failures deterministic across all environments we ignore known
+                // paths when calculating the string length, while still including the full
+                // path in the output. This could result in some output being larger than the
+                // threshold, but it's better than having nondeterministic failures.
+                //
+                // The compiler emitting only excluded strings is addressed by adding a
+                // placeholder size for each excluded segment, which will eventually reach
+                // the configured threshold.
+                for pattern in exclude_from_len {
+                    let pattern_bytes = pattern.as_bytes();
+                    // We start matching `pattern_bytes - 1` into the previously loaded data,
+                    // to account for the fact a pattern might be included across multiple
+                    // `extend` calls. Starting from `- 1` avoids double-counting patterns.
+                    let matches = (&bytes[(old_len.saturating_sub(pattern_bytes.len() - 1))..])
+                        .windows(pattern_bytes.len())
+                        .filter(|window| window == &pattern_bytes)
+                        .count();
+                    *excluded_len += matches as isize
+                        * (EXCLUDED_PLACEHOLDER_LEN - pattern_bytes.len() as isize);
```

Comment:
I think the EXCLUDED_PLACEHOLDER_LEN is pretty confusing to me (why 32?) -- it seems like the effect here is that we're going to sometimes shrink and sometimes increase the excluded_len depending on the pattern_bytes? It's hard for me to conceptualize what the effect of that will be in the long run.

Maybe a better/simpler approach here is to set an upper bound on the excluded bytes (say, 2 MB, or so) which seems like it would more clearly express that while we do want to avoid different failures across different systems, we probably also want to avoid enormous outputs from rustc. I think the effect of the placeholder length here is probably pretty similar regardless (right?), just less obviously correct.

Reply:
The effect of this is that all paths will be considered 32 bytes long, regardless of whether they're shorter or longer. This should remove all path size differences between different machines.

I chose 32 bytes because that's the highest power of two that passes all the tests. Specifically for `src/test/ui/numeric/numeric-cast.rs`, all path lengths greater than 53 cause that test to fail. If you want the behavior to be more consistent with today I guess we could go with 48?

Reply:
Why do we need to subtract pattern_bytes from EXCLUDED_PLACEHOLDER_LEN though? It looks like we have a total match count (`matches`) and if we want to ensure that no matter what the pattern is we'd count it as 32 bytes, we'd just multiply by 32 (`EXCLUDED_PLACEHOLDER_LEN`) here, right?

32 is fine with this rationale, but I do think we should put a comment on its definition saying something like "Number of bytes any excluded segment of the input will be treated as", with perhaps a short note for why (cross-platform consistency).

Reply:
> Why do we need to subtract pattern_bytes from EXCLUDED_PLACEHOLDER_LEN though? It looks like we have a total match count (matches) and if we want to ensure that no matter what the pattern is we'd count it as 32 bytes, we'd just multiply by 32 (EXCLUDED_PLACEHOLDER_LEN) here, right?

No, that'd still cause nondeterminism (`excluded_len` is the number of bytes that should be ignored from the output size).

The `pattern` is the path on the local filesystem (such as `/home/ci/rust`), so if we want to make things deterministic we need to ignore the full length of the pattern when counting the string size (so `matches * pattern_bytes.len()`).

That works, but it would lead to OOM if the compiler emits only the patterns without any extra bytes, as the whole output size would be ignored. To avoid that the code subtracts 32 from the pattern length: when the length of the pattern is lower than 32 the output will be considered *longer* (to simulate a system with longer paths), while when the length of the pattern is greater than 32 bytes it will be considered *shorter* (removing the nondeterminism, while preventing OOMs due to the 32 bytes still being counted).

In practice that's `32 - pattern_bytes` rather than `pattern_bytes - 32` due to the code below doing `len + excluded_bytes` rather than `len - excluded_bytes`.

I can change the code to invert signs (making things more clear) and add another comment explaining this.

Reply:
Ah, I think part of my confusion was that pattern is a local file system path (rather than something more abstract like a regex, e.g., "/[^"]+)

I think we should maybe consider renaming it to something like path_prefix since it isn't really a pattern - or, at least, doesn't match more than one specific str.

Reply:
I think the best thing here is to subtract `matches * pattern_bytes.len()` and then add `matches * 32`, which should I think be equivalent and much more obvious (to me, anyway), in terms of effect.

Reply:
Pushed a commit addressing this.